File: 03-herencia/python/demo03.py
"""
DEMO 03 - HERENCIA
Ejecutar:  python3 demo03.py

Cuatro archivos, igual que en Java:
    vehiculo.py   el padre
    auto.py       hijo
    moto.py       hijo
    demo03.py     el programa que los usa   <- estas aqui
"""

from auto import Auto
from moto import Moto
from vehiculo import Vehiculo


if __name__ == "__main__":

    auto = Auto("Toyota Yaris", 2020, 4)
    moto = Moto("Honda CB190", 2022, True)

    print("--- Metodos heredados del padre ---")
    auto.acelerar()
    moto.acelerar()

    print("--- Metodo sobrescrito en cada hijo ---")
    auto.describir()
    moto.describir()

    print("--- Metodos propios de cada hijo ---")
    auto.abrir_maletero()
    moto.hacer_caballito()

    print("--- Un hijo puede usarse como si fuera el padre ---")
    v = Auto("Ford Fiesta", 1998, 4)
    v.describir()
    v.acelerar()

    print("--- Lo que Python tiene y Java no ---")
    # En Python se puede preguntar el parentesco en tiempo de ejecucion
    print("¿auto es un Auto?    ", isinstance(auto, Auto))
    print("¿auto es un Vehiculo?", isinstance(auto, Vehiculo))
    print("¿auto es una Moto?   ", isinstance(auto, Moto))
    print("Cadena de herencia:  ", [c.__name__ for c in Auto.__mro__])

File: 03-herencia/java/Demo03.java
/*
 * DEMO 03 - HERENCIA (extends, super, @Override, protected)
 *
 * Compilar y ejecutar:
 *     javac *.java
 *     java Demo03
 *
 * Cuatro archivos, tres clases y un programa:
 *     Vehiculo.java   el padre
 *     Auto.java       hijo
 *     Moto.java       hijo
 *     Demo03.java     el programa que los usa   <- estas aqui
 */
public class Demo03 {
    public static void main(String[] args) {

        Auto auto = new Auto("Toyota Yaris", 2020, 4);
        Moto moto = new Moto("Honda CB190", 2022, true);

        System.out.println("--- Metodos heredados del padre ---");
        auto.acelerar();     // definido en Vehiculo.java, nadie lo reescribio
        moto.acelerar();

        System.out.println("--- Metodo sobrescrito en cada hijo ---");
        auto.describir();    // cada clase lo hace a su manera
        moto.describir();

        System.out.println("--- Metodos propios de cada hijo ---");
        auto.abrirMaletero();
        moto.hacerCaballito();

        System.out.println("--- Un hijo puede usarse como si fuera el padre ---");
        Vehiculo v = new Auto("Ford Fiesta", 1998, 4);
        v.describir();       // ejecuta la version de Auto, no la de Vehiculo
        v.acelerar();
        // v.abrirMaletero();  // ERROR: la variable es Vehiculo, no conoce ese metodo
    }
}
